web/src/engine/predictive-text/worker-thread/src/main/predict-helpers.ts

```diff
@@ -180,21 +181,51 @@ export async function correctAndEnumerate(
   // facilitates a more thorough correction-search pattern.
 
   // Token replacement benefits greatly from knowledge of the prior context state.
-  let contextState = contextTracker.analyzeState(
-    lexicalModel,
-    context,
-    null
-  ).final;
+  // Note:  not long-term viable - context is a sliding window, after all!
+  // What we actually need is a sliding-window comparison.
+  //
+  // Note that the "final" context from the last operation will not substitute any
+  // characters - only insert (if context window was shortened) or delete (if
+  // lengthened).  No substitutions possible, as no rules will have occurred - the
+  // ONLY change is the amount of known text vs the context window's range.
+  const lengthThreshold = contextTracker.configuration?.leftContextCodePoints ?? Number.POSITIVE_INFINITY;
+  const contextsMatch = (a: Context, b: Context) => {
+    // If either context's window is equal to or greater than the threshold length, there's a good
+    // chance something slid into or out of range.
+    if(a.left.length >= lengthThreshold || b.left.length >= lengthThreshold) {
+      return isSubstitutionAlignable(a.left, b.left);
+    } else {
+      // If both are smaller than the threshold, the text should match exactly.
+      return a.left == b.left;
+    }
+
+  };
+
+  const lastTransition = contextTracker.latest;
+  let contextState: ContextState;
+  if(contextsMatch(lastTransition.final.context, context)) {
+    contextState = lastTransition.final;
+  } else if(contextsMatch(lastTransition.base.context, context)) {
+    // Multitap case:  if we reverted back to the original underlying context,
+    // rather than using the previous output context.
+    contextState = lastTransition.base;
+  }
+
+  if(!contextState){
+    console.warn("Unexpected context state occurred as prediction base context");
+    contextTracker.reset(context, inputTransform.id);
+    contextState = contextTracker.latest.base;
+  }
 
   // Corrections and predictions are based upon the post-context state, though.
-  const contextChangeAnalysis = contextTracker.analyzeState(
-    lexicalModel,
-    context,
-    !TransformUtils.isEmpty(inputTransform)
-      ? transformDistribution
-      : null
-  );
-  const postContextState = contextChangeAnalysis.final;
+  let transition = ContextTracker.attemptMatchContext(context, lexicalModel, contextState, transformDistribution);
+  if(!transition) {
+    console.warn("Unexpected failure when computing context-state transition");
```

Comment:
While it would help us to log the details here... the most obvious details are user text, and we generally don't want to log that.  I'm currently trying to think up strategies or info that might help for a related Sentry event without adding major privacy concerns.